pandas/core/indexes/numeric.py

```diff
@@ -364,3 +368,137 @@ def _format_native_types(
+            np.uint16: libindex.UInt16Engine,
+            np.uint32: libindex.UInt32Engine,
+            np.uint64: libindex.UInt64Engine,
+            np.float32: libindex.Float32Engine,
+            np.float64: libindex.Float64Engine,
+        }[self.dtype.type]
+
+    @property
+    def inferred_type(self) -> str:
+        return {
+            "i": "integer",
+            "u": "integer",
+            "f": "floating",
+        }[self.dtype.kind]
+
+    @classmethod
+    def _ensure_array(cls, data, dtype, copy: bool) -> np.ndarray:
+        """
+        Ensure we have a valid array to pass to _simple_new.
+        """
+        arr = np.array(data, dtype=dtype, copy=copy)
+        cls._validate_dtype(arr.dtype)
+        return arr
+
+    @classmethod
+    def _assert_safe_casting(cls, data, subarr):
+        """
+        Ensure incoming data can be represented with matching signed-ness.
+        """
+        if data.dtype.kind not in {"i", "u"} and not np.array_equal(data, subarr):
+            raise TypeError("Unsafe NumPy casting, you must explicitly cast")
+
+    def __contains__(self, key) -> bool:
+        """
+        Check if key is a float and has a decimal. If it has, return False.
+        """
+        if not is_integer_dtype(self.dtype):
+            return super().__contains__(key)
+
+        hash(key)
+        try:
+            if is_float(key) and int(key) != key:
+                # otherwise the `key in self._engine` check casts e.g. 1.1 -> 1
+                return False
+            return key in self._engine
+        except (OverflowError, TypeError, ValueError):
+            return False
+
+    @doc(Index.astype)
+    def astype(self, dtype, copy=True):
+        if is_categorical_dtype(dtype):
+            from pandas import CategoricalIndex
+
+            return CategoricalIndex(self, name=self.name, dtype=dtype, copy=copy)
+        elif is_float_dtype(dtype):
+            dtype = pandas_dtype(dtype)
+            if needs_i8_conversion(dtype):
+                raise TypeError(
+                    f"Cannot convert Float64Index to dtype {dtype}; integer "
+                    "values are required for conversion"
+                )
+            elif is_integer_dtype(dtype) and not is_extension_array_dtype(dtype):
+                # TODO(jreback); this can change once we have an EA Index type
+                # GH 13149
+                arr = astype_nansafe(self._values, dtype=dtype)
+                return type(self)(arr, name=self.name, dtype=dtype)
+
+        return super().astype(dtype, copy=copy)
+
+    @doc(Index._convert_slice_indexer)
+    def _convert_slice_indexer(self, key: slice, kind: str):
+        assert kind in ["loc", "getitem"]
+
+        # We always treat __getitem__ slicing as label-based
+        # translate to locations
+        return self.slice_indexer(key.start, key.stop, key.step, kind=kind)
+
+    # ----------------------------------------------------------------
+
+    def _format_native_types(
```

Comment:
could this go in NumericIndex?  (i think could then remove from Float64Index?)

Reply:
Yes, done.